test/extensions/filters/network/redis_proxy/conn_pool_impl_test.cc

```diff
@@ -408,6 +425,55 @@ TEST_F(RedisConnPoolImplTest, ClientRequestFailed) {
   tls_.shutdownThread();
 };
 
+TEST_F(RedisConnPoolImplTest, RedisConnectionRateLimited) {
+  InSequence s;
+
+  setup(true, true, 100, nullptr, 1);
+
+  Common::Redis::RespValue value;
+  Common::Redis::Client::MockPoolRequest active_request;
+  MockPoolCallbacks callbacks;
+  Common::Redis::Client::MockClient* client = new NiceMock<Common::Redis::Client::MockClient>();
+  EXPECT_CALL(cm_.thread_local_cluster_.lb_, chooseHost(_))
+      .WillOnce(Invoke([&](Upstream::LoadBalancerContext* context) -> Upstream::HostConstSharedPtr {
+        EXPECT_EQ(context->computeHashKey().value(), MurmurHash::murmurHash2("hash_key"));
+        EXPECT_EQ(context->metadataMatchCriteria(), nullptr);
+        EXPECT_EQ(context->downstreamConnection(), nullptr);
+        return cm_.thread_local_cluster_.lb_.host_;
+      }));
+  EXPECT_CALL(*this, create_(_)).WillOnce(Return(client));
+  EXPECT_CALL(*cm_.thread_local_cluster_.lb_.host_, address())
+      .WillRepeatedly(Return(test_address_));
+  EXPECT_CALL(*client, makeRequest_(Eq(value), _)).WillOnce(Return(&active_request));
+  Common::Redis::Client::PoolRequest* request =
+      conn_pool_->makeRequest("hash_key", ConnPool::RespVariant(value), callbacks, transaction_);
+  EXPECT_NE(nullptr, request);
+  EXPECT_EQ(connectionRateLimited().value(), 0);
+
+  // close local and reconnect, should be rate limited and get null
+  EXPECT_CALL(tls_.dispatcher_, deferredDelete_(_));
+  client->raiseEvent(Network::ConnectionEvent::LocalClose);
+
+  EXPECT_CALL(cm_.thread_local_cluster_.lb_, chooseHost(_))
+      .WillOnce(Invoke([&](Upstream::LoadBalancerContext* context) -> Upstream::HostConstSharedPtr {
+        EXPECT_EQ(context->computeHashKey().value(), MurmurHash::murmurHash2("hash_key"));
+        EXPECT_EQ(context->metadataMatchCriteria(), nullptr);
+        EXPECT_EQ(context->downstreamConnection(), nullptr);
+        return cm_.thread_local_cluster_.lb_.host_;
+      }));
+  EXPECT_CALL(*this, create_(_)).Times(0);
+  EXPECT_CALL(*cm_.thread_local_cluster_.lb_.host_, address())
+      .WillRepeatedly(Return(test_address_));
+  Common::Redis::Client::PoolRequest* request2 =
+      conn_pool_->makeRequest("hash_key", ConnPool::RespVariant(value), callbacks, transaction_);
+  EXPECT_EQ(nullptr, request2);
+  EXPECT_EQ(connectionRateLimited().value(), 1);
+
```

Comment:
What about a test for recovering from the rate limited state?

Reply:
Seems we don't allow using `sleep_for()` for testings so looks like I need to inject a `timesystem` somehow to the conn_pool_impl? Not sure if we want to add a field injection just for this but please suggest. (sry that I am still learning the practice of the project :)

Reply:
I think you should be using a mock TLS and mock dispatcher here, so time should be mockable. Probably no need for additional dependency injection.

Reply:
Looks good now, thanks.